### PR TITLE
[UXIT-1588] Allocators Filters

### PR DESCRIPTION
**Pull request by CharlyMartin**
## 📝 Description

This PR adds two select filters to the allocator table: type and region.

I originally tried to use the table's internal to implement it, but that approach didn’t work out as expected. So, I switched to using React's `useState` to manage the filter and sort states.

## 🛠️ Key Changes

1. Stopped using `setGlobalFilter` in favor of column filters: the search input now filters only the “name” column
2. Created `SelectFilter` to manage type and location filters state
3. Added `AllocatorsTableFiltersLayout` to handle the positioning of the search and select inputs
4. Created `getColumnOrThrow` util to throw an error if `table.getColumn` returns `undefined`. This util function follows the same principle as `findOrThrow` to make the code simpler.
5. Created `createUniqueOptionsFromData` to dynamically build a list of options from the allocators' data

## 🧪 How to Test

1. Go to `/filecoin-plus/allocators`
2. Test the search input, it should only search through the allocators' name
3. Test the region and type filters

## 📸 Screenshots

CleanShot 2024-10-28 at 10 29 19

CleanShot 2024-10-28 at 10 29 45

## 🔖 Resources

- https://tanstack.com/table/v8/docs/guide/column-filtering
- https://tanstack.com/table/latest/docs/api/core/table#getcolumn

### CODE REVIEW

File: src/app/filecoin-plus/allocators/components/AllocatorsTableWithFilters.tsx [hunk 1]
```diff
@@ -22,30 +30,55 @@ type AllocatorsTableWithFiltersProps = {
...
+          />
+        }
+        locationFilter={
+          <SelectFilter
+            column={locationColumn}
+            options={locationOptions}
+            resetOptionName="All Regions"
+          />
+        }
+      />
```
Comment (mirhamasala) [inline src/app/filecoin-plus/allocators/components/AllocatorsTableWithFilters.tsx:81]: From [>>](https://github.com/FilecoinFoundationWeb/filecoin-foundation/pull/754/files#r1819025319)

```suggestion
          <AllocatorsTableFiltersLayout data={data} table={table} />
```

File: src/app/filecoin-plus/allocators/components/SelectFilter.tsx [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  name: string
+}
+
+type SortListboxProps = {
+  column: Column<Allocator>
+  options: ReadonlyArray<FilterOption>
+  resetOptionName: string
+}
+
+const EMPTY_FILTER_VALUE = ''
```
Comment (mirhamasala) [inline src/app/filecoin-plus/allocators/components/SelectFilter.tsx:27]:
```suggestion
const DEFAULT_FILTER_ID = ''
```

File: src/app/filecoin-plus/allocators/components/SelectFilter.tsx [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  options,
+  resetOptionName,
+}: SortListboxProps) {
+  const currentFilterValue =
+    column.getFilterValue()?.toString() || EMPTY_FILTER_VALUE
+
+  const resetOption: FilterOption = {
+    id: EMPTY_FILTER_VALUE,
+    name: resetOptionName,
+  }
```
Comment (mirhamasala) [inline src/app/filecoin-plus/allocators/components/SelectFilter.tsx:40]: `resetOption` and `resetOptionName` throws me off a bit. What about something like..

```suggestion
    const defaultOption: FilterOption = {
      id: DEFAULT_FILTER_ID,
      name: defaultOptionLabel,
    }
```

And then

```
<SelectFilter
  column={typeColumn}
  options={typeOptions}
  defaultOptionLabel="All Types"
/>
```

File: src/app/filecoin-plus/allocators/components/SelectFilter.tsx [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+const EMPTY_FILTER_VALUE = ''
+
+export function SelectFilter({
+  column,
+  options,
+  resetOptionName,
+}: SortListboxProps) {
+  const currentFilterValue =
+    column.getFilterValue()?.toString() || EMPTY_FILTER_VALUE
```
Comment (mirhamasala) [inline src/app/filecoin-plus/allocators/components/SelectFilter.tsx:35]: Do we need `|| EMPTY_FILTER_VALUE`?

File: src/app/filecoin-plus/allocators/utils/createUniqueOptionsFromData.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+function countOccurrences(array: Array<string>) {
+  return array.reduce(
+    (object, item) => {
+      const curentCount = object[item] || 0
+      object[item] = curentCount + 1
+      return object
+    },
+    {} as Record<string, number>,
+  )
+}
```
Comment (mirhamasala) [inline src/app/filecoin-plus/allocators/utils/createUniqueOptionsFromData.ts:34]:
```suggestion
  const valuesArray = data.map((item) => item[key])
  const valueCounts = countOccurrences(valuesArray)

  const sortedUniqueOptions = Object.keys(valueCounts)
    .sort((a, b) => valueCounts[b] - valueCounts[a])
    .map((value) => ({
      id: value,
      name: value,
    }))

  return sortedUniqueOptions
}

function countOccurrences(array: Array<string>) {
  return array.reduce((counts, item) => {
    counts[item] = (counts[item] || 0) + 1
    return counts
  }, {} as Record<string, number>)
}
```

Comment (mirhamasala) [reply src/app/filecoin-plus/allocators/utils/createUniqueOptionsFromData.ts:34]: Sidenote - I'm happy to keep `reduce` but I do think it tends to throw off people. `forEach` can do the same thing.

```
function countOccurrences(array: Array<string>) {
  const counts: Record<string, number> = {}

  array.forEach((item) => {
    if (counts[item]) {
      counts[item] += 1
    } else {
      counts[item] = 1
    }
  })

  return counts
}
```

Comment (mirhamasala) [reply src/app/filecoin-plus/allocators/utils/createUniqueOptionsFromData.ts:34]: Lastly, what about `generateSortedOptions`?

File: src/app/filecoin-plus/allocators/components/AllocatorsTableFiltersLayout.tsx [hunk 6]
```diff
@@ -0,0 +1,24 @@
+import type { ReactNode } from 'react'
+
+type AllocatorsTableFiltersLayoutProps = {
+  nameSearch: ReactNode
+  typeFilter: ReactNode
+  locationFilter: ReactNode
```
[the remark this answers is not in the document]
Comment (CharlyMartin) [reply src/app/filecoin-plus/allocators/components/AllocatorsTableFiltersLayout.tsx:6]: I used the same logic as in `FilterContainer` and created a stateless container component in order to keep the state in the parent.

File: src/app/filecoin-plus/allocators/components/SelectFilter.tsx (hunk 4 above)
Comment (CharlyMartin) [reply src/app/filecoin-plus/allocators/components/SelectFilter.tsx:35]: The `toString()` part is to make TypeScript happy as the return type of `getFilterValue` is `unknown`.

https://tanstack.com/table/v8/docs/api/features/column-filtering#getfiltervalue

Comment (CharlyMartin) [reply src/app/filecoin-plus/allocators/components/SelectFilter.tsx:35]: But it looks like we don't need it anymore in this scenario! 👌

File: src/app/filecoin-plus/allocators/components/SelectFilter.tsx (hunk 3 above)
Comment (CharlyMartin) [reply src/app/filecoin-plus/allocators/components/SelectFilter.tsx:40]: I'm happy with the renaming, good suggestions. Not a big fan of of declaring a `DEFAULT_OPTION` constant and then overwriting it with a prop.

I've reworked the component, let me know what you think of this new version :)

File: src/app/filecoin-plus/allocators/utils/createUniqueOptionsFromData.ts (hunk 5 above)
Comment (CharlyMartin) [reply src/app/filecoin-plus/allocators/utils/createUniqueOptionsFromData.ts:34]: Love it :)

File: src/app/filecoin-plus/allocators/components/AllocatorsTableWithFilters.tsx (hunk 1 above)
Comment (CharlyMartin) [reply src/app/filecoin-plus/allocators/components/AllocatorsTableWithFilters.tsx]: I reworked it as per your suggestion

**Comment by CharlyMartin** [comment]
Thanks for the suggestions @mirhamasala!

- I've used more descriptive names in `SelectFilter`
- I've reworked `generateSortedOptions` as you suggested
- Finally, I've encapsulated the filter components in `AllocatorsTableFilters`

**Review by mirhamasala — APPROVED** [review]
@CharlyMartin - Thank you for the rework. This looks great! ❤️🚀

**Pull request merged**